File: src/faebryk/core/zig/zigpilled/src/sexp/performance_sexp.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.8"
# dependencies = [
#     "sexpdata",
# ]
# ///

import subprocess
import time
from pathlib import Path

import sexpdata

TEST_FILES_DIR = Path(__file__).parent / "test_files"
ZIG_OUT_DIR = Path(__file__).parent / "zig-out" / "bin"
PERFORMANCE_TEST_EXE = ZIG_OUT_DIR / "performance_sexp"


def count_tokens(data, depth=0):
    """Recursively count tokens and S-expressions in parsed data"""
    lparen = 0
    rparen = 0
    symbols = 0
    numbers = 0
    strings = 0
    lists = 0
    atoms = 0

    def count_recursive(item):
        nonlocal lparen, rparen, symbols, numbers, strings, lists, atoms

        if isinstance(item, list):
            lparen += 1
            rparen += 1
            lists += 1
            for sub_item in item:
                count_recursive(sub_item)
        elif isinstance(item, sexpdata.Symbol):
            symbols += 1
            atoms += 1
        elif isinstance(item, (int, float)):
            numbers += 1
            atoms += 1
        elif isinstance(item, str):
            strings += 1
            atoms += 1

    count_recursive(data)
    return (
        lparen,
        rparen,
        symbols,
        numbers,
        strings,
        0,
        lists,
        atoms,
    )  # 0 comments - Python parser strips them


def test_python(file_path):
    """Test Python sexpdata performance"""
    # Read file
    file_contents = file_path.read_text()
    file_size = len(file_contents)

    # Parse with timing
    start_time = time.time() * 1000  # milliseconds
    try:
        data = sexpdata.loads(file_contents)
        end_time = time.time() * 1000
        parse_time = int(end_time - start_time)

        # Count tokens
        lparen, rparen, symbols, numbers, strings, comments, lists, atoms = (
            count_tokens(data)
        )
        total_tokens = lparen + rparen + symbols + numbers + strings + comments

        return {
            "file_size": file_size,
            "tokenize_time": parse_time,  # Python does both in one pass
            "parse_time": 0,  # Already included in tokenize_time
            "total_tokens": total_tokens,
            "lparen": lparen,
            "rparen": rparen,
            "symbols": symbols,
            "numbers": numbers,
            "strings": strings,
            "comments": comments,
            "balanced": lparen == rparen,
            "sexp_count": 1 if isinstance(data, list) else 0,
            "list_count": lists,
            "atom_count": atoms,
            "error": None,
        }
    except Exception as e:
        return {"file_size": file_size, "error": str(e)}


def compile_zig_if_needed():
    """Build the Zig performance test using the build system"""
    build_file = Path(__file__).parent / "build.zig"

    if not build_file.exists():
        return {"error": f"Build file {build_file} not found"}

    # Check if we need to rebuild
    source_files = list(Path(__file__).parent.glob("*.zig"))
    if not source_files:
        return {"error": "No Zig source files found"}

    needs_compile = not PERFORMANCE_TEST_EXE.exists()
    if not needs_compile:
        # Check if any source is newer than executable
        exe_mtime = PERFORMANCE_TEST_EXE.stat().st_mtime
        for source in source_files:
            if source.stat().st_mtime > exe_mtime:
                needs_compile = True
                break

    if needs_compile:
        print("Building Zig performance test...")
        result = subprocess.run(
            ["zig", "build", "-Doptimize=ReleaseFast", "--build-file", str(build_file)],
            cwd=Path(__file__).parent,
            capture_output=True,
            text=True,
        )
        if result.returncode != 0:
            return {"error": f"Failed to build: {result.stderr}"}
        print("Build successful")

    return None


def test_zig(file_path):
    """Test Zig parallel tokenizer and AST parser performance"""
    # Compile if needed
    compile_error = compile_zig_if_needed()
    if compile_error:
        return compile_error

    # Run the performance test
    result = subprocess.run(
        [str(PERFORMANCE_TEST_EXE), str(file_path.absolute())],
        capture_output=True,
        text=True,
    )

    if result.returncode != 0:
        return {"error": f"Zig test failed: {result.stderr}"}

    # Parse output
    for line in result.stderr.split("\n"):
        if line.startswith("RESULT:"):
            parts = line.replace("RESULT:", "").split(":")
            if len(parts) >= 15:
                return {
                    "file_size": int(parts[0]),
                    "tokenize_time": int(parts[1]),
                    "parse_time": int(parts[2]),
                    "total_tokens": int(parts[3]),
                    "lparen": int(parts[4]),
                    "rparen": int(parts[5]),
                    "symbols": int(parts[6]),
                    "numbers": int(parts[7]),
                    "strings": int(parts[8]),
                    "comments": int(parts[9]),
                    "balanced": parts[10] == "true",
                    "sexp_count": int(parts[11]),
                    "list_count": int(parts[12]),
                    "atom_count": int(parts[13]),
                    "cpu_count": int(parts[14]),
                    "error": None,
                }

    return {"error": "Could not parse Zig output"}


def format_results(name, results):
    """Format test results for display"""
    if results.get("error"):
        return f"{name}: ERROR - {results['error']}"

    tokens_str = (
        f"(L:{results['lparen']} R:{results['rparen']} "
        f"S:{results['symbols']} N:{results['numbers']} "
        f"Str:{results['strings']}"
    )

    # Add comments if present
    if results.get("comments", 0) > 0:
        tokens_str += f" C:{results['comments']}"

    tokens_str += ")"

    # Build timing string
    if results.get("parse_time", 0) > 0:
        time_str = (
            f"tokenize:{results['tokenize_time']}ms "
            f"+ parse:{results['parse_time']}ms = "
            f"{results['tokenize_time'] + results['parse_time']}ms"
        )
    else:
        time_str = f"{results['tokenize_time']}ms"

    # Add S-expression counts if available
    sexp_str = ""
    if "list_count" in results and "atom_count" in results:
        sexp_str = (
            f", S-exps: {results.get('sexp_count', 0)}"
            f" (Lists:{results['list_count']} Atoms:{results['atom_count']})"
        )

    return (
        f"{name}: {results['file_size']} bytes, "
        f"{time_str}, "
        f"{results['total_tokens']} tokens "
        f"{tokens_str} "
        f"{'✓' if results['balanced'] else '✗'}"
        f"{sexp_str}"
    )


def main():
    # Find all .kicad_pcb files in test_files directory
    test_files = sorted(TEST_FILES_DIR.glob("*.kicad_pcb"))

    if not test_files:
        print("No .kicad_pcb files found in test_files directory")
        return

    print(f"Found {len(test_files)} test files\n")

    for test_file in test_files:
        print(f"=== {test_file.name} ===")

        # Test Python
        python_results = test_python(test_file)
        print(format_results("Python", python_results))

        # Test Zig
        zig_results = test_zig(test_file)
        zig_name = "Zig"
        if zig_results.get("cpu_count"):
            zig_name += f" ({zig_results['cpu_count']} cores)"
        print(format_results(zig_name, zig_results))

        # Show speedup if both succeeded
        if not python_results.get("error") and not zig_results.get("error"):
            python_time = python_results["tokenize_time"] + python_results.get(
                "parse_time", 0
            )
            zig_time = zig_results["tokenize_time"] + zig_results["parse_time"]

            if zig_time > 0:
                speedup = python_time / zig_time
                print(f"Speedup: {speedup:.1f}x")
            else:
                print("Speedup: Too fast to measure accurately")

        # Show throughput
        if not zig_results.get("error"):
            file_size_mb = zig_results["file_size"] / (1024 * 1024)
            total_time_sec = (
                zig_results["tokenize_time"] + zig_results["parse_time"]
            ) / 1000.0
            if total_time_sec > 0:
                throughput = file_size_mb / total_time_sec
                print(f"Zig throughput: {throughput:.1f} MB/s")

        print()


if __name__ == "__main__":
    main()

File: src/faebryk/core/zig/zigpilled/src/sexp/performance_sexp.zig
const std = @import("std");
const tokenizer = @import("tokenizer.zig");
const ast = @import("ast.zig");
const netlist = @import("kicad/netlist.zig");

const TokenCounts = struct {
    lparen: usize = 0,
    rparen: usize = 0,
    symbols: usize = 0,
    numbers: usize = 0,
    strings: usize = 0,
    comments: usize = 0,
};

fn countTokenTypes(tokens: []const tokenizer.Token) TokenCounts {
    var counts = TokenCounts{};

    for (tokens) |token| {
        switch (token.type) {
            .lparen => counts.lparen += 1,
            .rparen => counts.rparen += 1,
            .symbol => counts.symbols += 1,
            .number => counts.numbers += 1,
            .string => counts.strings += 1,
            .comment => counts.comments += 1,
        }
    }

    return counts;
}

const SExpCounts = struct {
    lists: usize = 0,
    atoms: usize = 0,
};

fn countSExpElements(sexp: ?ast.SExp) SExpCounts {
    var counts = SExpCounts{};

    if (sexp) |s| {
        countSExpRecursive(s, &counts);
    }

    return counts;
}

fn countSExpRecursive(sexp: ast.SExp, counts: *SExpCounts) void {
    switch (sexp.value) {
        .list => |items| {
            counts.lists += 1;
            for (items) |item| {
                countSExpRecursive(item, counts);
            }
        },
        .symbol, .number, .string, .comment => counts.atoms += 1,
    }
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);

    if (args.len != 2) {
        std.debug.print("Usage: {s} <file.kicad_pcb>\n", .{args[0]});
        std.process.exit(1);
    }

    const file_path = args[1];

    // Get CPU count
    const cpu_count = try std.Thread.getCpuCount();
    std.debug.print("CPU count: {}\n", .{cpu_count});

    var timer = try std.time.Timer.start();

    // Read file size
    const file = try std.fs.cwd().openFile(file_path, .{});
    const file_size = try file.getEndPos();
    file.close();

    std.debug.print("File size: {} bytes ({d:.2} MB)\n", .{ file_size, @as(f64, @floatFromInt(file_size)) / (1024.0 * 1024.0) });
    std.debug.print("\n", .{});

    // Read file content
    const file_content = try std.fs.cwd().readFileAlloc(allocator, file_path, 100 * 1024 * 1024);
    defer allocator.free(file_content);
    const read_time = timer.read();
    std.debug.print("Read time: {d:.3} ms\n", .{@as(f64, @floatFromInt(read_time)) / 1_000_000.0});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(read_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});

    // Test 1: Sequential tokenization
    std.debug.print("=== Sequential Tokenization ===\n", .{});

    timer.reset();
    const tokens_seq = try tokenizer._tokenize(allocator, file_content);
    defer {
        allocator.free(tokens_seq);
    } // No need for deinitTokens - tokens point to file_content

    const seq_time = timer.read();
    std.debug.print("Time: {d:.3} ms\n", .{@as(f64, @floatFromInt(seq_time)) / 1_000_000.0});
    std.debug.print("Tokens: {}\n", .{tokens_seq.len});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(seq_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});
    std.debug.print("\n", .{});

    // Test 2: File tokenization (with duplication)
    std.debug.print("=== Parallel Tokenization ===\n", .{});
    timer.reset();

    const tokens_par = try tokenizer.tokenize(allocator, file_content);
    defer allocator.free(tokens_par);

    const par_time = timer.read();
    std.debug.print("Time: {d:.3} ms\n", .{@as(f64, @floatFromInt(par_time)) / 1_000_000.0});
    std.debug.print("Tokens: {}\n", .{tokens_par.len});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(par_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});
    std.debug.print("Speedup: {d:.2}x\n", .{@as(f64, @floatFromInt(seq_time)) / @as(f64, @floatFromInt(par_time))});
    std.debug.print("\n", .{});

    // Test 3: Full parsing with arena allocator
    std.debug.print("=== Full Parsing (Arena Allocator) ===\n", .{});

    var arena = std.heap.ArenaAllocator.init(allocator);
    defer arena.deinit();

    timer.reset();
    var sexp = try ast.parse(arena.allocator(), tokens_seq);
    defer sexp.deinit(arena.allocator());
    const parse_time = timer.read();

    timer.reset();
    var structure_time: u64 = 0;
    // Only perform structure decoding if the file is a .net file
    if (std.mem.endsWith(u8, file_path, ".net")) {
        // The parsed sexp should be the (export ...) expression
        // Remove debug output and just pass it directly

        var netlistfile = try netlist.NetlistFile.loads(arena.allocator(), .{ .sexp = sexp });
        defer netlistfile.free(arena.allocator());
        structure_time = timer.read();
    }

    std.debug.print("Tokenize time: {d:.3} ms\n", .{@as(f64, @floatFromInt(par_time)) / 1_000_000.0});
    std.debug.print("Parse time: {d:.3} ms\n", .{@as(f64, @floatFromInt(parse_time)) / 1_000_000.0});
    std.debug.print("Structure time: {d:.3} ms\n", .{@as(f64, @floatFromInt(structure_time)) / 1_000_000.0});
    std.debug.print("Total time (tokenize + parse + structure): {d:.3} ms\n", .{@as(f64, @floatFromInt(par_time + parse_time + structure_time)) / 1_000_000.0});
    std.debug.print("Speed: {d:.2} MB/s\n", .{@as(f64, @floatFromInt(file_size)) / (@as(f64, @floatFromInt(par_time + parse_time + structure_time)) / 1_000_000_000.0) / (1024.0 * 1024.0)});

    // Verify results match
    std.debug.print("\n=== Verification ===\n", .{});
    if (tokens_seq.len != tokens_par.len) {
        std.debug.print("WARNING: Token counts don't match! Sequential: {}, File: {}\n", .{ tokens_seq.len, tokens_par.len });
    } else {
        std.debug.print("✓ Token counts match\n", .{});
    }

    // Output in format expected by Python script
    const counts = countTokenTypes(tokens_seq);
    const sexp_counts = countSExpElements(sexp);
    const tokenize_time_ms = @divFloor(seq_time, 1_000_000);
    const parse_time_ms = @divFloor(parse_time, 1_000_000);
    const structure_time_ms = @divFloor(structure_time, 1_000_000);

    // RESULT:file_size:tokenize_time:parse_time:total_tokens:lparen:rparen:symbols:numbers:strings:comments:balanced:sexp_count:list_count:atom_count:cpu_count
    std.debug.print("RESULT:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}:{}\n", .{
        file_size,
        tokenize_time_ms,
        parse_time_ms,
        structure_time_ms,
        tokens_seq.len,
        counts.lparen,
        counts.rparen,
        counts.symbols,
        counts.numbers,
        counts.strings,
        counts.comments,
        counts.lparen == counts.rparen,
        //if (sexp != null) @as(usize, 1) else @as(usize, 0),
        @as(usize, 1),
        sexp_counts.lists,
        sexp_counts.atoms,
        cpu_count,
    });
}
